casper/src/main/scala/coop/rchain/casper/engine/Running.scala

```diff
@@ -284,107 +250,115 @@ class Running[F[_]: Sync: BlockStore: CommUtil: TransportLayer: ConnectionsCell:
-        F.whenA(b.sender == self)(
-          Log[F].warn(
-            s"There is another node $peer proposing using the same private key as you. Or did you restart your node?"
-          )
-        )
-      } >> {
-      for {
-        maybeReq <- RequestedBlocks.get(b.blockHash)
-        _ <- maybeReq match {
-              // There might be blocks that are not maintained by RequestedBlocks, e.g. genesis ceremony messages
-              case None =>
-                Log[F].info(
-                  s"Block ${PrettyPrinter.buildString(b.blockHash)} is not present in RequestedBlocks."
+  private def ignoreCasperMessage(hash: BlockHash): F[IgnoreCasperMessageStatus] =
+    BlocksRetriever[F]
+      .received(hash)
+      .ifM(
+        IgnoreCasperMessageStatus(doIgnore = true, BlockIsWaitingForCasper).pure[F],
+        CasperBufferStorage[F]
+          .contains(hash)
+          .ifM(
+            IgnoreCasperMessageStatus(doIgnore = true, BlockIsInCasperBuffer).pure[F],
+            casper.blockDag.flatMap(
+              _.contains(hash)
+                .ifM(
+                  IgnoreCasperMessageStatus(doIgnore = true, BlockIsInDag).pure[F],
+                  IgnoreCasperMessageStatus(doIgnore = false, DoNotIgnore).pure[F]
                 )
-              case Some(requested) => {
-                // Make Casper loop aware that the block has been received
-                RequestedBlocks.put(b.blockHash, requested.copy(received = true)) >>
-                  Log[F].info(
-                    s"Block ${PrettyPrinter.buildString(b.blockHash)} marked as received."
-                  )
-              }
-            }
-      } yield ()
-    } >> processUnfinishedParentsAndAddBlock(b)
-  }
+            )
+          )
+      )
 
   /**
-    * Add parents that are already in block store and then block itself
-    * @param b block
-    * @return Result of block processing
+    * Basic block check before saving block in Blockstore. The intention here is to prevent saving
+    * bad and useless blocks. So the check is if block is well formed and of interest for this particular node.
+    * Should not catch any slashable offences - such blocks should be persisted and processed by Casper.
+    * @param b BlockMessage
+    * @param peer peer from which BlockMessage came from
+    * @return if block is suitable for this node
     */
-  private def processUnfinishedParentsAndAddBlock(b: BlockMessage): F[ValidBlockProcessing] =
-    processUnfinishedParents(b) >> casper.addBlock(b)
+  private def blockIsOk(
+      b: BlockMessage,
+      peer: PeerNode
+  ): F[Boolean] =
+    withCasper(
+      casper => {
+        validatorId match {
+          case None =>
+          case Some(id) => {
+            val self = ByteString.copyFrom(id.publicKey.bytes)
+            val _ = F.whenA(b.sender == self)(
+              Log[F].warn(
+                s"There is another node $peer proposing using the same private key as you. Or did you restart your node?"
```

Comment:
This expression is also "in the air", Log.warn will never be executed. Instead of `Unit` it should return `F[Unit]` and executed as part of for comprehension bellow.